tests/codegen/option-nonzero-eq.rs

```diff
@@ -37,8 +37,15 @@ pub fn non_null_eq(l: Option<NonNull<u8>>, r: Option<NonNull<u8>>) -> bool {
 // CHECK-lABEL: @ordering_eq
 #[no_mangle]
 pub fn ordering_eq(l: Option<Ordering>, r: Option<Ordering>) -> bool {
-    // CHECK: start:
-    // CHECK-NEXT: icmp eq i8
-    // CHECK-NEXT: ret i1
+    // CHECK-NOT: ret i1
+    // CHECK: %[[A1:.+]] = icmp ult i8
+    // CHECK-NEXT: call void @llvm.assume(i1 %[[A1]])
+    // CHECK-NOT: ret i1
+    // CHECK: %[[A2:.+]] = icmp ult i8
+    // CHECK-NEXT: call void @llvm.assume(i1 %[[A2]])
+    // CHECK-NOT: ret i1
+    // CHECK: %[[CMP:.+]] = icmp eq i8 %0, %1
+    // CHECK-NEXT: ret i1 %[[CMP]]
+    // CHECK-NOT: ret i1
     l == r
 }
```

Comment:
Without the library change, this was giving
```llvm
define noundef zeroext i1 @ordering_eq(i8 noundef %0, i8 noundef %1) unnamed_addr #0 personality ptr @__CxxFrameHandler3 {
start:
  %2 = icmp eq i8 %0, 2
  br i1 %2, label %"_ZN4core6option15Option$LT$T$GT$6map_or17h10b30776faeff97fE.exit", label %bb3.i

bb3.i:                                            ; preds = %start
  %3 = add i8 %0, 1
  %_7.i.i = icmp ult i8 %3, 3
  tail call void @llvm.assume(i1 %_7.i.i)
  br label %"_ZN4core6option15Option$LT$T$GT$6map_or17h10b30776faeff97fE.exit"

"_ZN4core6option15Option$LT$T$GT$6map_or17h10b30776faeff97fE.exit": ; preds = %start, %bb3.i
  %4 = icmp eq i8 %1, 2
  br i1 %4, label %"_ZN4core6option15Option$LT$T$GT$6map_or17hd4add697368b47d3E.exit", label %bb3.i2

bb3.i2:                                           ; preds = %"_ZN4core6option15Option$LT$T$GT$6map_or17h10b30776faeff97fE.exit"
  %5 = add i8 %1, 1
  %_7.i.i1 = icmp ult i8 %5, 3
  tail call void @llvm.assume(i1 %_7.i.i1)
  br label %"_ZN4core6option15Option$LT$T$GT$6map_or17hd4add697368b47d3E.exit"

"_ZN4core6option15Option$LT$T$GT$6map_or17hd4add697368b47d3E.exit": ; preds = %"_ZN4core6option15Option$LT$T$GT$6map_or17h10b30776faeff97fE.exit", %bb3.i2
  %6 = icmp eq i8 %0, %1
  ret i1 %6
}
```

Now that it's using the `transmute` instead of the `map_or` approach, it's
```llvm
define noundef zeroext i1 @ordering_eq(i8 noundef %0, i8 noundef %1) unnamed_addr #0 {
start:
  %2 = add i8 %0, 1
  %3 = icmp ult i8 %2, 4
  tail call void @llvm.assume(i1 %3)
  %4 = add i8 %1, 1
  %5 = icmp ult i8 %4, 4
  tail call void @llvm.assume(i1 %5)
  %6 = icmp eq i8 %0, %1
  ret i1 %6
}
```

Reply:
While we can do this hack in libcore, this means user code will regress in similar situations. Should we track this issue to figure out how to improve it? We could autogenerate the transmute in a MIR opt for example.

Reply:
Hmm, actually, maybe it's not a *real* problem.  It looks bad in the *LLVM* IR, but the assembly generated for it is still the correct single test: <https://rust.godbolt.org/z/da819G99e>
```asm
example::qux:
        cmp     dil, sil
        sete    al
        ret
```
despite
```llvm
define noundef zeroext i1 @_ZN7example3qux17hb5cd12f675ddd331E(i8 noundef %x, i8 noundef %y) unnamed_addr #0 personality ptr @rust_eh_personality {
  %0 = icmp eq i8 %x, -128
  br i1 %0, label %"_ZN4core6option15Option$LT$T$GT$6map_or17h8cb56c5fe57f6f5dE.exit", label %bb3.i

bb3.i:                                            ; preds = %start
  %_4.i.i = icmp sgt i8 %x, -1
  tail call void @llvm.assume(i1 %_4.i.i)
  br label %"_ZN4core6option15Option$LT$T$GT$6map_or17h8cb56c5fe57f6f5dE.exit"

"_ZN4core6option15Option$LT$T$GT$6map_or17h8cb56c5fe57f6f5dE.exit": ; preds = %start, %bb3.i
  %1 = icmp eq i8 %y, -128
  br i1 %1, label %"_ZN4core6option15Option$LT$T$GT$6map_or17h93b3ae047b8aab61E.exit", label %bb3.i2

bb3.i2:                                           ; preds = %"_ZN4core6option15Option$LT$T$GT$6map_or17h8cb56c5fe57f6f5dE.exit"
  %_4.i.i1 = icmp sgt i8 %y, -1
  tail call void @llvm.assume(i1 %_4.i.i1)
  br label %"_ZN4core6option15Option$LT$T$GT$6map_or17h93b3ae047b8aab61E.exit"

"_ZN4core6option15Option$LT$T$GT$6map_or17h93b3ae047b8aab61E.exit": ; preds = %"_ZN4core6option15Option$LT$T$GT$6map_or17h8cb56c5fe57f6f5dE.exit", %bb3.i2
  %2 = icmp eq i8 %x, %y
  ret i1 %2
}
```

So I suppose I could also try moving these tests to assembly tests if that would be better, and leaving the `map_or`s.

Reply:
Based on the thumb, I'll do that.

@rustbot author

Reply:
Ok, removed the `library/` change, and instead moved the test to assembly instead.  Didn't change the logic at all.

@rustbot ready